orderer/common/cluster/deliver.go

```diff
@@ -114,11 +114,18 @@ func (p *BlockPuller) HeightsByEndpoints() (map[string]uint64, error) {
 
 func (p *BlockPuller) tryFetchBlock(seq uint64) *common.Block {
 	var reConnected bool
-	for p.isDisconnected() {
+
+	for retriesLeft := p.MaxPullBlockRetries; p.isDisconnected(); {
```

Comment:
I don't think we need to attempt to pop if we exhausted the attempt count, because when we try to pull any block and fail doing so, we will anyway drop the entire buffer (when we return from **pullBlocks** with an error we close the puller and it resets the buffer), and whenever we close the puller we anyway reset the buffer. 

So, why not just return nil? I also argue that whenever we pull a sequence of blocks and keep them in the buffer, subsequent invocations of **tryFetchBlock** won't even try to connect anyway. It is only until the buffer is full, when we would detect we are disconnected.

Also let's move the retriesLeft-- to the for loop and do something like:

```

	for retriesLeft := p.MaxPullBlockRetries; p.isDisconnected(); retriesLeft-- {
		reConnected = true
		p.connectToSomeEndpoint(seq)
		if p.isDisconnected() {
			time.Sleep(p.RetryTimeout)
		}

		if p.MaxPullBlockRetries > 0 && retriesLeft <= 0 {
			p.Logger.Errorf("Failed connecting to some endpoint, attempts exhausted(%d)", p.MaxPullBlockRetries)
			return nil
		}
	}
```